File: packages/python/taiwan_payroll/engine/supplementary.py
from __future__ import annotations

from .._rounding import apply_rate, not_finite_number
from .._types import SupplementaryInput, SupplementaryResult


def calc_supplementary(data: dict, inp: SupplementaryInput, rounding: str) -> SupplementaryResult:
    sp = data["supplementaryPremium"]
    if not_finite_number(inp.amount) or inp.amount < 0:
        raise ValueError(f"amount must be a finite non-negative number, got {inp.amount}")

    if inp.type == "bonus":
        mis = inp.monthly_insured_salary
        if mis is None or not_finite_number(mis) or mis <= 0:
            raise ValueError(f"bonus requires a positive monthlyInsuredSalary, got {mis}")
        ytd = inp.ytd_bonus if inp.ytd_bonus is not None else 0
        if not_finite_number(ytd) or ytd < 0:
            raise ValueError(f"ytdBonus must be a finite non-negative number, got {ytd}")
        threshold = sp["bonusThresholdMultiplier"] * mis
        chargeable = int(max(0, ytd + inp.amount - max(ytd, threshold)))
    else:
        threshold = data["minimumWage"]["monthly"] if inp.type == "parttime" else sp["lowerThreshold"]
        chargeable = int(min(inp.amount, sp["singlePaymentCap"])) if inp.amount >= threshold else 0

    return SupplementaryResult(
        type=inp.type,
        chargeable=chargeable,
        rate=sp["rate"],
        premium=apply_rate(chargeable, [sp["rate"]], rounding),
    )


def calc_dividend_premium(data: dict, amount: float, employer_insured_total: float = 0) -> int:
    """股利扣繳補充保費（一般股東/雇主常見情況）。

    一般股東：單次給付 × 費率（單次達下限起扣）；雇主：(單次給付 − 投保額總額) × 費率。
    股票股利/特殊註記等情形不在涵蓋範圍，請由呼叫端自行判定後以 record.premium 提供。
    """
    sp = data["supplementaryPremium"]
    if amount < sp["lowerThreshold"]:
        return 0
    base = int(max(0, min(amount, sp["singlePaymentCap"]) - employer_insured_total))
    return apply_rate(base, [sp["rate"]], "round")

File: packages/core/src/engine/supplementary.ts
import type { YearData, Rounding, SupplementaryInput, SupplementaryResult } from '../types';
import { applyRate } from '../rounding/strategies';

/**
 * 二代健保補充保費 (2.11%). premium = round(chargeable × rate).
 *  - bonus: threshold = bonusThresholdMultiplier × this payment's current-month insured;
 *           chargeable = the part of THIS payment above the running cumulative threshold (uses ytdBonus).
 *  - parttime: threshold = minimum wage; others: threshold = lowerThreshold (20,000).
 *           non-bonus charges the FULL amount (capped at singlePaymentCap) when amount >= threshold.
 * Source: 健保署「補充保險費計算公式」cp-4516.
 */
export function calcSupplementary(
  data: YearData,
  input: SupplementaryInput,
  rounding: Rounding,
): SupplementaryResult {
  const sp = data.supplementaryPremium;
  if (!Number.isFinite(input.amount) || input.amount < 0) {
    throw new Error(`amount must be a finite non-negative number, got ${input.amount}`);
  }

  let chargeable: number;
  if (input.type === 'bonus') {
    if (
      input.monthlyInsuredSalary === undefined ||
      !Number.isFinite(input.monthlyInsuredSalary) ||
      input.monthlyInsuredSalary <= 0
    ) {
      throw new Error(`bonus requires a positive monthlyInsuredSalary, got ${input.monthlyInsuredSalary}`);
    }
    const ytd = input.ytdBonus ?? 0;
    if (!Number.isFinite(ytd) || ytd < 0) {
      throw new Error(`ytdBonus must be a finite non-negative number, got ${ytd}`);
    }
    const threshold = sp.bonusThresholdMultiplier * input.monthlyInsuredSalary;
    chargeable = Math.max(0, ytd + input.amount - Math.max(ytd, threshold));
  } else {
    const threshold = input.type === 'parttime' ? data.minimumWage.monthly : sp.lowerThreshold;
    chargeable = input.amount >= threshold ? Math.min(input.amount, sp.singlePaymentCap) : 0;
  }

  return { type: input.type, chargeable, rate: sp.rate, premium: applyRate(chargeable, [sp.rate], rounding) };
}

/**
 * 便利函式：算「一般股東/雇主」常見情況的股利扣繳補充保險費。
 *  - 一般股東 (employerInsuredTotal = 0)：min(amount, singlePaymentCap) × rate
 *  - 雇主：(min(amount, singlePaymentCap) − 投保額總額) × rate
 * 單次給付未達 lowerThreshold (2 萬) 不扣。
 * 不涵蓋股票股利/特殊註記等情形——那些由呼叫端自行判定後以 record.premium 提供。
 */
export function calcDividendPremium(data: YearData, amount: number, employerInsuredTotal = 0): number {
  const sp = data.supplementaryPremium;
  if (amount < sp.lowerThreshold) return 0;
  const base = Math.max(0, Math.min(amount, sp.singlePaymentCap) - employerInsuredTotal);
  return applyRate(base, [sp.rate], 'round');
}
